ops/jujuversion.py

```diff
@@ -0,0 +1,63 @@
+import re
+from functools import total_ordering
+
+
+@total_ordering
+class JujuVersion:
+
+    PATTERN = r'^(?P<major>\d{1,9})\.(?P<minor>\d{1,9})((?:\.|-(?P<tag>[a-z]+))(?P<patch>\d{1,9}))?(\.(?P<build>\d{1,9}))?$'
+
+    def __init__(self, version):
+        m = re.match(self.PATTERN, version)
+        if not m:
+            raise RuntimeError(f"{version} is not a valid Juju version string.")
```

Comment:
No dots at the end of error messages.  The version string should also be quoted in this context so that spaces and other awkward characters that break it may be read properly.